executor/src/witgen/query_processor.rs

```diff
@@ -65,105 +70,95 @@ where
-                    .iter()
-                    .find_map(|MatchArm { pattern, value }| match pattern {
-                        MatchPattern::CatchAll => Some(Ok(value)),
-                        MatchPattern::Pattern(pattern) => {
-                            match self.evaluate_expression(pattern, rows) {
-                                Ok(p) => (p == v).then_some(Ok(value)),
-                                Err(e) => Some(Err(e)),
-                            }
-                        }
-                    })
-                    .ok_or(IncompleteCause::NoMatchArmFound)??;
-                self.interpolate_query(expr, rows)
-            }
-            _ => self.evaluate_expression(query, rows).map(|v| v.to_string()),
-        }
+    ) -> Result<String, EvalError> {
+        let arguments = vec![rows.current_row_index.into()];
+        evaluator::evaluate_function_call(
+            query,
+            arguments,
+            &Symbols {
+                fixed_data: self.fixed_data,
+                rows,
+            },
+        )
+        .map(|v| v.to_string())
     }
+}
 
-    fn evaluate_expression(
-        &self,
-        expr: &'a Expression<T>,
-        rows: &RowPair<T>,
-    ) -> Result<T, IncompleteCause<&'a AlgebraicReference>> {
-        match expr {
-            Expression::Number(n) => Ok(*n),
-            Expression::BinaryOperation(left, op, right) => Ok(evaluate_binary_operation(
-                self.evaluate_expression(left, rows)?,
-                *op,
-                self.evaluate_expression(right, rows)?,
-            )),
-            Expression::UnaryOperation(op, inner) => Ok(evaluate_unary_operation(
-                *op,
-                self.evaluate_expression(inner, rows)?,
-            )),
-            Expression::Reference(Reference::LocalVar(i, _name)) => {
-                assert!(*i == 0);
-                Ok(rows.current_row_index.into())
-            }
-            Expression::FunctionCall(FunctionCall { id, arguments }) => {
-                let [arg] = &arguments[..] else {
-                    return Err(IncompleteCause::ExpressionEvaluationUnimplemented(
-                        expr.to_string(),
-                    ));
-                };
-                let arg = self.evaluate_expression(arg, rows)?;
-                self.evaluate_poly(id.to_string().as_str(), arg.to_degree(), rows)
-            }
-            e => Err(IncompleteCause::ExpressionEvaluationUnimplemented(
-                e.to_string(),
-            )),
-        }
-    }
+#[derive(Clone)]
+struct Symbols<'a, T: FieldElement> {
+    // TODO we should also provide access to non-column symbols.
+    fixed_data: &'a FixedData<'a, T>,
+    rows: &'a RowPair<'a, 'a, T>,
+}
 
-    fn evaluate_poly(
+impl<'a, T: FieldElement> SymbolLookup<'a, T, Reference> for Symbols<'a, T> {
+    fn lookup(&self, name: &str) -> Result<Value<'a, T, Reference>, EvalError> {
+        let poly_id = self.fixed_data.column_by_name(name);
```

Comment:
This panics if the column doesn't exist, I guess instead we want to return a `EvalError::SymbolNotFound`?